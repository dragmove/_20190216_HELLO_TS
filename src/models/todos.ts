import { types, getRoot, destroy, flow } from 'mobx-state-tree';
import { SHOW_ALL, SHOW_COMPLETED, SHOW_ACTIVE } from '../constants/TodoFilters';

// TODO: What is types.union, types.literal.
const filterType = types.union(...[SHOW_ALL, SHOW_COMPLETED, SHOW_ACTIVE].map(types.literal));
const TODO_FILTERS = {
  [SHOW_ALL]: () => true,
  [SHOW_COMPLETED]: (todo: any) => !todo.completed,
  [SHOW_ACTIVE]: (todo: any) => todo.completed,
};

const Todo = types
  .model('Todo', {
    text: types.string, // no default value
    completed: types.optional(types.boolean, false), // default value is false.
    id: types.identifierNumber, // no default value
  })
  .actions(self => ({
    // getRoot(self) is TodoStore
    remove() {
      const root: any = getRoot(self);
      root.removeTodo(self);
    },
    edit(text: string) {
      self.text = text;
    },
    complete() {
      self.completed = !self.completed;
    },
  }));

const TodoStore = types
  .model('TodoStore', {
    todos: types.array(Todo),
    filter: types.optional(filterType, SHOW_ALL),
  })
  .views((self: any) => ({
    // "computed" values

    // self is TodoStore
    get completedCount(): number {
      return self.todos.reduce((count: number, todo: any) => (todo.completed ? count + 1 : count), 0);
    },
    get activeCount(): number {
      return self.todos.length - self.completedCount;
    },
    get filteredTodos() {
      return self.todos.filter(TODO_FILTERS[self.filter]);
    },
    // can define Model views functions here.
  }))
  .actions(self => {
    // Use generator. fetch data.
    const callAPI = flow(function*() {
      // do something

      try {
        const res: any = yield window.fetch('https://jsonplaceholder.typicode.com/todos/1');
        const data: any = yield res.json();
        console.log('[callAPI] success. data :', data);

        // success. do something by data
      } catch (error) {
        // fail. error
        console.log('[callAPI] fail. error :', error);
      }

      // do something
    });

    return {
      addTodo(text: string) {
        const id = self.todos.reduce((maxId, todo) => Math.max(todo.id, maxId), -1) + 1;
        self.todos.unshift({
          id,
          text,
        });
      },
      removeTodo(todo: any) {
        destroy(todo);
      },
      completeAll() {
        const areAllMarked = self.todos.every(todo => todo.completed);
        self.todos.forEach(todo => (todo.completed = !areAllMarked));
      },
      clearCompleted() {
        self.todos.replace(self.todos.filter(todo => todo.completed === false));
      },
      setFilter(filter: string) {
        self.filter = filter;
      },

      callAPI,
      /*
      // Use async, await. fetch data.
      async callAPI() {
        const url: string = 'https://jsonplaceholder.typicode.com/todos/1';

        const res: any = await window.fetch(url);
        console.log('[callAPI] res :', res);

        if (res.ok) {
          const data = res.json();
          console.log('[callAPI] data :', data);

          // success. do something by data
        }

        // fail. error
      },
      */
    };
  })
  .preProcessSnapshot(snapshot => {
    console.log('TodoStore preProcessSnapshot :', snapshot);
    return snapshot;
  })
  .postProcessSnapshot(snapshot => {
    console.log('TodoStore postProcessSnapshot :', snapshot);
    return snapshot;
  });

export default TodoStore;
