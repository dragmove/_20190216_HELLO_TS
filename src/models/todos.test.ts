import { getSnapshot, onPatch } from 'mobx-state-tree';
import { Todo, TodoStore } from './todos';

it('can create a instance of Todo model', () => {
  const todo = Todo.create({
    text: 'test mobx-state-tree',
    id: 1,
  });

  // test model values
  expect(todo.completed).toBe(false);
  expect(todo.text).toBe('test mobx-state-tree');
  expect(todo.id).toBe(1);

  // test changes by call actions
  const patches: any[] = [];
  onPatch(todo, patch => patches.push(patch));

  todo.edit('test complete');
  expect(todo.text).toBe('test complete');

  /*
  // can change from
  expect(getSnapshot(todo)).toEqual({
    text: 'test complete',
    completed: false,
    id: 1,
  });
  // to
  expect(getSnapshot(todo)).toMatchSnapshot();
  */

  todo.complete();
  expect(todo.completed).toBe(true);

  expect(patches).toMatchSnapshot();
});

it('can create a instance of TodoStore model', () => {
  const snapshot = {
    todos: [
      {
        text: 'learn Mobx',
        completed: false,
        id: 0,
      },
      {
        text: 'learn MST',
        completed: false,
        id: 1,
      },
    ],
  };

  const todoStore = TodoStore.create(snapshot);

  expect(todoStore.todos.length).toBe(2);
  expect(todoStore.todos[0].text).toBe('learn Mobx');
  expect(todoStore.todos[0].id).toBe(0);
});
