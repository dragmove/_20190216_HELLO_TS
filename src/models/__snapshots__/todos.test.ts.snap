// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a instance of Todo model 1`] = `
Array [
  Object {
    "op": "replace",
    "path": "/text",
    "value": "test complete",
  },
  Object {
    "op": "replace",
    "path": "/completed",
    "value": true,
  },
]
`;
